File: cli/bun/src/core/service/triage.ts
// `cli service triage`: one read that answers "what state is
// this session in and what should I run next?". Mirrors
// cli/rust/crates/prose-runner-core/src/service/triage.rs byte for byte.
//
// It composes GET /health, /wallet/balance, /organizations/default,
// /runs?limit=5 and /triggers. Every section carries
// its own `problem`, so one failing read never hides the others and the
// operation still exits 0; only cancellation (and invocation errors) end it.
// Without a usable credential the account sections are null and `credential`
// names the exact variable to set. Price-side fields only.
// Result: shared/schemas/service/discovery.schema.json#/$defs/serviceTriage.
// Decisions: docs/service/discovery.md.
import { failure } from "../errors";
import { humanSafeScalar } from "../output";
import { RunnerFailure } from "../types";
import { projectHealth, Shape } from "./discovery";
import { jsonObject, requestFor, type Response } from "./http";
import type { Context } from "./index";
import { projectJobs } from "./jobs";
import type { Json, JsonObject } from "./manifest";
import { organization } from "./organizations";
import { argvText, redactDetails } from "./render";
import { projectRun } from "./run-records";
import { projectBalance } from "./wallet";

/** Run statuses that mean the run is still going (`cli run watch` applies). */
export const LIVE_RUN_STATUSES: readonly string[] = ["queued", "pending", "starting", "running", "input_needed", "awaiting_input"];
/** The low-balance threshold of the wallet section and its top-up suggestion. */
export const DEFAULT_LOW_BALANCE_CENTS = 500;
export const RECENT_RUNS = 5;
export const JOBS_SHOWN = 5;
export const LIVE_WATCH_MAX = 3;
export const FAILING_JOBS_MAX = 2;
const JOB_KEYS = ["id", "type", "name", "next_fire_at", "next_fire_at_iso", "last_event_at", "last_event_at_iso", "last_run_id", "last_error"] as const;
const HEALTH_KEYS = ["status", "default_model", "models"] as const;

// Indices into the manifest's service.triage requests.
const HEALTH = 0;
const WALLET = 1;
const ORGANIZATION = 2;
const RUNS = 3;
const JOBS = 4;

function protocol(reason: string): RunnerFailure {
  return failure("SERVICE_PROTOCOL_INVALID", { reason });
}

async function fetch(context: Context, index: number, path: string): Promise<Response> {
  const request = requestFor(context.operation, index, path);
  if (index === RUNS) request.query.push(["limit", String(RECENT_RUNS)]);
  return await context.send(request);
}

/** Runs one read; a failure is returned (cancellation and non-runner errors are thrown). */
async function attempt(read: () => Promise<JsonObject>): Promise<JsonObject | RunnerFailure> {
  try { return await read(); }
  catch (caught) {
    if (caught instanceof RunnerFailure && caught.code !== "CANCELLED") return caught;
    throw caught;
  }
}

function problemJson(error: RunnerFailure): Json { return error.toJSON() as unknown as Json; }

/** A section: the projected fields plus `problem: null`, or `{problem}`. */
function section(context: Context, projected: JsonObject | RunnerFailure): JsonObject {
  if (projected instanceof RunnerFailure) {
    redactDetails(projected.details as Record<string, unknown> | undefined, context.knownCredential());
    return { problem: problemJson(projected) };
  }
  return { ...projected, problem: null };
}

/** A failed /health with no serviceStatus: nothing answered, skip the other reads. */
function unreachable(error: RunnerFailure): boolean {
  return error.code === "SERVICE_UNAVAILABLE" && !Object.hasOwn(error.details ?? {}, "serviceStatus");
}

function detail(error: RunnerFailure, key: string): string | undefined {
  const value = (error.details ?? {})[key];
  return typeof value === "string" ? value : undefined;
}

export async function execute(context: Context): Promise<Json> {
  const healthRead = await attempt(async () => {
    const health = projectHealth(new Shape("service status").object(await fetch(context, HEALTH, "/health")));
    const kept: JsonObject = {};
    for (const key of HEALTH_KEYS) if (Object.hasOwn(health, key)) kept[key] = health[key]!;
    return kept;
  });
  const reachable = !(healthRead instanceof RunnerFailure && unreachable(healthRead));
  const health = section(context, healthRead);

  const variable = context.environment.credentialEnv;
  let state: string;
  let source: string;
  let credentialProblem: RunnerFailure | undefined;
  try {
    await context.credentialValue();
    state = "unverified";
    source = context.credentialOrigin() ?? "environment";
  } catch (caught) {
    if (!(caught instanceof RunnerFailure) || caught.code === "CANCELLED") throw caught;
    state = caught.code === "CREDENTIAL_STORE_UNAVAILABLE" ? "unavailable" : detail(caught, "credentialProblem") === "malformed" ? "malformed" : "missing";
    const origin = detail(caught, "credentialSource");
    source = origin === "environment" || origin === "store" ? origin : "none";
    redactDetails(caught.details as Record<string, unknown> | undefined, context.knownCredential());
    credentialProblem = caught;
  }

  const account: Json[] = [null, null, null, null];
  if (reachable && credentialProblem === undefined) {
    const reads: Array<() => Promise<JsonObject>> = [
      () => walletSection(context),
      () => organizationSection(context),
      () => runsSection(context),
      () => jobsSection(context),
    ];
    for (const [slot, read] of reads.entries()) {
      const projected = await attempt(read);
      if (!(projected instanceof RunnerFailure)) {
        state = "valid";
        account[slot] = section(context, projected);
        continue;
      }
      const annotated = context.annotateAuth(projected);
      if (detail(annotated, "credentialProblem") === "rejected") {
        redactDetails(annotated.details as Record<string, unknown> | undefined, context.knownCredential());
        state = "rejected";
        credentialProblem = annotated;
        break;
      }
      account[slot] = section(context, annotated);
    }
  }
  const [wallet, org, runs, jobs] = account;
  const result: JsonObject = {
    health,
    credential: { variable, source, state, problem: credentialProblem === undefined ? null : problemJson(credentialProblem) },
    wallet: wallet!,
    organization: org!,
    runs: runs!,
    jobs: jobs!,
  };
  result.nextCommands = nextCommands(context, result, reachable);
  context.human = human(result);
  return result;
}

async function walletSection(context: Context): Promise<JsonObject> {
  const body = jsonObject(await fetch(context, WALLET, "/wallet/balance"));
  const balance = projectBalance(body);
  const threshold = DEFAULT_LOW_BALANCE_CENTS;
  const available = typeof balance.available_cents === "number" ? balance.available_cents : 0;
  return { balance, low: available < threshold, lowBelowCents: threshold };
}

async function organizationSection(context: Context): Promise<JsonObject> {
  const body = jsonObject(await fetch(context, ORGANIZATION, "/organizations/default"));
  return { default: organization(Object.hasOwn(body, "organization") ? body.organization : undefined) };
}

/** The run fields `service triage` reports (user fields only). */
const TRIAGE_RUN_FIELDS = ["run_id", "status", "model", "created_at", "program_ref", "price_cents"];

/** The user fields of a recent run: triage never carries runtime details. */
function triageRun(run: JsonObject): JsonObject {
  return Object.fromEntries(Object.entries(run).filter(([key]) => TRIAGE_RUN_FIELDS.includes(key)));
}

async function runsSection(context: Context): Promise<JsonObject> {
  const body = jsonObject(await fetch(context, RUNS, "/runs"));
  const rows = body.runs;
  if (!Array.isArray(rows) || rows.length > 200) throw protocol("the run list is missing or has more than 200 runs");
  const recent = rows.slice(0, RECENT_RUNS).map((run) => triageRun(projectRun(run, false)));
  const live = recent.filter((run) => typeof run.status === "string" && LIVE_RUN_STATUSES.includes(run.status)).map((run) => run.run_id!);
  return { recent, live };
}

async function jobsSection(context: Context): Promise<JsonObject> {
  const body = jsonObject(await fetch(context, JOBS, "/triggers"));
  const { jobs: all, max } = projectJobs(body);
  const shown = all.slice(0, JOBS_SHOWN).map((entry) => {
    const job: JsonObject = {};
    for (const key of JOB_KEYS) {
      const value = entry[key];
      if (value !== undefined && value !== null) job[key] = value;
    }
    return job;
  });
  return { total: all.length, max, jobs: shown };
}

function objectOf(value: Json | undefined): JsonObject {
  return value !== null && typeof value === "object" && !Array.isArray(value) ? value : {};
}

function reasonOf(problem: Json | undefined): string {
  const shape = objectOf(problem);
  const reason = objectOf(shape.details).reason;
  if (typeof reason === "string") return reason;
  return typeof shape.message === "string" ? shape.message : "";
}

function centsText(cents: number): string {
  return `${Math.trunc(cents / 100)}.${String(Math.abs(cents % 100)).padStart(2, "0")}`;
}

/** nextCommands in a fixed order: credential, unreachable, live runs, low balance, failing jobs. */
function nextCommands(context: Context, result: JsonObject, reachable: boolean): Json[] {
  const next: Json[] = [];
  const credential = objectOf(result.credential);
  if (credential.problem !== null && credential.problem !== undefined) {
    next.push({ why: reasonOf(credential.problem), argv: null, env: credential.variable! });
  }
  if (!reachable) {
    next.push({ why: "the service did not respond; check your network connection, then retry", argv: context.followUpArgv(["service", "status"]), env: null });
  }
  const runs = objectOf(result.runs);
  if (Array.isArray(runs.live)) {
    const recent = Array.isArray(runs.recent) ? runs.recent.map(objectOf) : [];
    for (const runId of runs.live.filter((value): value is string => typeof value === "string").slice(0, LIVE_WATCH_MAX)) {
      const status = recent.find((run) => run.run_id === runId)?.status;
      next.push({ why: `run ${runId} is ${typeof status === "string" ? status : ""}`, argv: context.followUpArgv(["run", "watch", runId]), env: null });
    }
  }
  const wallet = objectOf(result.wallet);
  if (wallet.low === true) {
    const threshold = typeof wallet.lowBelowCents === "number" ? wallet.lowBelowCents : 0;
    const available = objectOf(wallet.balance).available_dollars;
    next.push({
      why: `the available balance $${typeof available === "string" ? available : ""} is below $${centsText(threshold)}`,
      argv: context.followUpArgv(["wallet", "topup", "--amount-cents", String(threshold), "--preview"]),
      env: null,
    });
  }
  const jobs = objectOf(result.jobs);
  if (Array.isArray(jobs.jobs)) {
    for (const job of jobs.jobs.map(objectOf).filter((entry) => Object.hasOwn(entry, "last_error")).slice(0, FAILING_JOBS_MAX)) {
      const id = typeof job.id === "string" ? job.id : "";
      next.push({ why: `job ${id} reported an error`, argv: context.followUpArgv(["job", "show", id]), env: null });
    }
  }
  return next;
}

function text(value: Json | undefined): string { return humanSafeScalar(typeof value === "string" ? value : ""); }

function problemLine(label: string, problem: Json | undefined): string {
  const code = objectOf(problem).code;
  return `${label}: ${typeof code === "string" ? code : ""}: ${humanSafeScalar(reasonOf(problem))}\n`;
}

/** The human report: at most 25 lines. Exported for the unit test. */
export function human(result: JsonObject): string {
  let out = "";
  const health = objectOf(result.health);
  if (health.problem === null) {
    out += `Service: ${text(health.status)}, default model ${text(health.default_model)}\n`;
  } else out += problemLine("Service", health.problem);
  const credential = objectOf(result.credential);
  const variable = text(credential.variable);
  const origin =
    credential.source === "none"
      ? `${variable} not set`
      : credential.source === "environment"
        ? `from ${variable}`
        : credential.source === "store"
          ? "from the credential store"
          : typeof credential.source === "string"
            ? `${variable} from ${credential.source}`
            : variable;
  out += `Credential: ${text(credential.state)}, ${origin}`;
  out += credential.problem === null ? "\n" : `; ${humanSafeScalar(reasonOf(credential.problem))}\n`;
  if (result.wallet === null) out += "Wallet: skipped\n";
  else {
    const wallet = objectOf(result.wallet);
    if (wallet.problem === null) {
      const balance = objectOf(wallet.balance);
      out += `Wallet: $${text(balance.available_dollars)} available, $${text(balance.reserved_dollars)} reserved`;
      if (wallet.low === true) out += ` (low: below $${centsText(typeof wallet.lowBelowCents === "number" ? wallet.lowBelowCents : 0)})`;
      out += "\n";
    } else out += problemLine("Wallet", wallet.problem);
  }
  if (result.organization === null) out += "Organization: skipped\n";
  else {
    const org = objectOf(result.organization);
    if (org.problem === null) {
      const entry = objectOf(org.default);
      out += `Organization: ${text(entry.slug)}`;
      if (typeof entry.role === "string") out += ` (${humanSafeScalar(entry.role)})`;
      out += "\n";
    } else out += problemLine("Organization", org.problem);
  }
  if (result.runs === null) out += "Runs: skipped\n";
  else {
    const runs = objectOf(result.runs);
    if (runs.problem === null) {
      const recent = Array.isArray(runs.recent) ? runs.recent.map(objectOf) : [];
      const live = Array.isArray(runs.live) ? runs.live.length : 0;
      out += `Runs: ${recent.length} recent, ${live} live\n`;
      for (const run of recent) out += `  ${text(run.run_id)}  ${text(run.status)}  ${text(run.model)}  ${text(run.created_at)}\n`;
    } else out += problemLine("Runs", runs.problem);
  }
  if (result.jobs === null) out += "Jobs: skipped\n";
  else {
    const jobs = objectOf(result.jobs);
    if (jobs.problem === null) {
      const total = typeof jobs.total === "number" ? jobs.total : 0;
      out += typeof jobs.max === "number" ? `Jobs: ${total} of ${jobs.max} allowed\n` : `Jobs: ${total}\n`;
      for (const job of Array.isArray(jobs.jobs) ? jobs.jobs.map(objectOf) : []) {
        const name = Object.hasOwn(job, "name") ? text(job.name) : "-";
        out += `  ${text(job.id)}  ${text(job.type)}  ${name}${Object.hasOwn(job, "last_error") ? "  (error)" : ""}\n`;
      }
    } else out += problemLine("Jobs", jobs.problem);
  }
  const next = Array.isArray(result.nextCommands) ? result.nextCommands.map(objectOf) : [];
  if (next.length === 0) out += "Next: nothing needs attention\n";
  else {
    out += "Next:\n";
    for (const command of next) {
      const action = Array.isArray(command.argv)
        ? argvText(command.argv.filter((word): word is string => typeof word === "string"))
        : `set ${typeof command.env === "string" ? command.env : ""}`;
      out += `  ${humanSafeScalar(action)}  # ${humanSafeScalar(typeof command.why === "string" ? command.why : "")}\n`;
    }
  }
  return out;
}

File: cli/rust/crates/prose-runner-core/src/service/triage.rs
//! `cli service triage`: one read that answers "what state
//! is this session in and what should I run next?".
//!
//! It composes the reads an agent otherwise makes one by one: `GET /health`,
//! `GET /wallet/balance`, `GET /organizations/default`, `GET /runs?limit=5`
//! and `GET /triggers`. Every section carries its own
//! `problem`, so one failing read never hides the others and the operation
//! still exits 0; only cancellation (and invocation errors) end it early.
//! Without a usable credential the account sections are `null` and the
//! `credential` section names the exact variable to set. Only price-side
//! fields are projected (the balance, run prices); nothing cost-side.
//! The result is `shared/schemas/service/discovery.schema.json#/$defs/serviceTriage`;
//! the Bun twin is `cli/bun/src/core/service/triage.ts`. Decisions:
//! docs/service/discovery.md.
use super::{Context, discovery, http, jobs, organizations, render, run_records, wallet};
use crate::RunnerError;
use crate::error::{ErrorCode, human_safe_scalar};
use serde_json::{Map, Value, json};
use std::fmt::Write as _;

/// Run statuses that mean the run is still going (`cli run watch` applies).
/// Any other status, including an unknown one, is treated as finished.
pub const LIVE_RUN_STATUSES: [&str; 6] = [
    "queued",
    "pending",
    "starting",
    "running",
    "input_needed",
    "awaiting_input",
];
/// The low-balance threshold.
pub const DEFAULT_LOW_BALANCE_CENTS: i64 = 500;
/// Recent runs requested and kept (the manifest's `limit`).
const RECENT_RUNS: usize = 5;
/// The run fields `service triage` reports (user fields only).
const TRIAGE_RUN_FIELDS: [&str; 6] = [
    "run_id",
    "status",
    "model",
    "created_at",
    "program_ref",
    "price_cents",
];
/// Jobs kept in the `jobs` section.
const JOBS_SHOWN: usize = 5;
/// `run watch` suggestions for live runs.
const LIVE_WATCH_MAX: usize = 3;
/// `job show` suggestions for jobs reporting an error.
const FAILING_JOBS_MAX: usize = 2;
/// The job fields a triage keeps (a subset of `cli job list`).
const JOB_KEYS: [&str; 9] = [
    "id",
    "type",
    "name",
    "next_fire_at",
    "next_fire_at_iso",
    "last_event_at",
    "last_event_at_iso",
    "last_run_id",
    "last_error",
];

/// The `cli service status` fields a triage keeps.
const HEALTH_KEYS: [&str; 3] = ["status", "default_model", "models"];

/// Indexes into the manifest's `service.triage` requests.
const HEALTH: usize = 0;
const WALLET: usize = 1;
const ORGANIZATION: usize = 2;
const RUNS: usize = 3;
const JOBS: usize = 4;

fn protocol(reason: &str) -> RunnerError {
    RunnerError::catalog(ErrorCode::ServiceProtocolInvalid).with_detail("reason", reason)
}

fn fetch(
    context: &mut Context<'_>,
    index: usize,
    path: &str,
) -> Result<http::Response, RunnerError> {
    let mut request = http::Request::from_manifest(context.operation, index, path);
    if index == RUNS {
        request = request.query("limit", RECENT_RUNS.to_string());
    }
    context.send(&request)
}

/// A section: the projected fields plus `problem: null`, or `{problem}`.
/// Cancellation is not a section problem: it ends the operation.
fn section(
    context: &Context<'_>,
    projected: Result<Map<String, Value>, RunnerError>,
) -> Result<Value, RunnerError> {
    match projected {
        Ok(mut fields) => {
            fields.insert("problem".into(), Value::Null);
            Ok(Value::Object(fields))
        }
        Err(error) if error.code == ErrorCode::Cancelled => Err(error),
        Err(mut error) => {
            render::redact_error(&mut error, context.known_credential());
            Ok(json!({"problem": error}))
        }
    }
}

/// Whether a failed `/health` means nothing answered at all (a transport
/// failure carries no `serviceStatus`); then the other reads are skipped.
fn unreachable(error: &RunnerError) -> bool {
    error.code == ErrorCode::ServiceUnavailable
        && error
            .details
            .as_ref()
            .is_none_or(|details| !details.contains_key("serviceStatus"))
}

fn detail<'e>(error: &'e RunnerError, key: &str) -> Option<&'e str> {
    error.details.as_ref()?.get(key)?.as_str()
}

/// Executes `cli service triage`.
pub fn execute(context: &mut Context<'_>) -> Result<Value, RunnerError> {
    let health = fetch(context, HEALTH, "/health").and_then(|response| {
        let mut health = discovery::health(&response)?;
        health.retain(|key, _| HEALTH_KEYS.contains(&key.as_str()));
        Ok(health)
    });
    let reachable = !matches!(&health, Err(error) if unreachable(error));
    let health = section(context, health)?;

    let variable = context.environment.credential_env;
    let (mut state, source, mut credential_problem) = match context.credential() {
        Ok(_) => (
            "unverified",
            context.credential_source.unwrap_or("environment"),
            None,
        ),
        Err(error) if error.code == ErrorCode::Cancelled => return Err(error),
        Err(mut error) => {
            let state = if error.code == ErrorCode::CredentialStoreUnavailable {
                "unavailable"
            } else {
                match detail(&error, "credentialProblem") {
                    Some("malformed") => "malformed",
                    _ => "missing",
                }
            };
            let source = match detail(&error, "credentialSource") {
                Some("environment") => "environment",
                Some("store") => "store",
                _ => "none",
            };
            render::redact_error(&mut error, context.known_credential());
            (state, source, Some(error))
        }
    };

    let mut account = [Value::Null, Value::Null, Value::Null, Value::Null];
    if reachable && credential_problem.is_none() {
        for (slot, index) in [WALLET, ORGANIZATION, RUNS, JOBS].into_iter().enumerate() {
            let projected = match index {
                WALLET => wallet_section(context),
                ORGANIZATION => organization_section(context),
                RUNS => runs_section(context),
                _ => jobs_section(context),
            };
            match projected {
                Ok(fields) => {
                    state = "valid";
                    account[slot] = section(context, Ok(fields))?;
                }
                Err(error) if error.code == ErrorCode::Cancelled => return Err(error),
                Err(mut error) => {
                    context.annotate_auth(&mut error);
                    if detail(&error, "credentialProblem") == Some("rejected") {
                        render::redact_error(&mut error, context.known_credential());
                        state = "rejected";
                        credential_problem = Some(error);
                        break;
                    }
                    account[slot] = section(context, Err(error))?;
                }
            }
        }
    }
    let [wallet, organization, runs, jobs] = account;
    let credential = json!({
        "variable": variable,
        "source": source,
        "state": state,
        "problem": credential_problem,
    });
    let mut result = json!({
        "health": health,
        "credential": credential,
        "wallet": wallet,
        "organization": organization,
        "runs": runs,
        "jobs": jobs,
    });
    let next = next_commands(context, &result, reachable);
    result["nextCommands"] = Value::Array(next);
    context.human = Some(human(&result));
    Ok(result)
}

fn wallet_section(context: &mut Context<'_>) -> Result<Map<String, Value>, RunnerError> {
    let body = fetch(context, WALLET, "/wallet/balance")?.json_object()?;
    let balance = wallet::project_balance(&body)?;
    let threshold = DEFAULT_LOW_BALANCE_CENTS;
    let available = balance["available_cents"].as_i64().unwrap_or_default();
    let mut fields = Map::new();
    fields.insert("balance".into(), balance);
    fields.insert("low".into(), json!(available < threshold));
    fields.insert("lowBelowCents".into(), json!(threshold));
    Ok(fields)
}

fn organization_section(context: &mut Context<'_>) -> Result<Map<String, Value>, RunnerError> {
    let body = fetch(context, ORGANIZATION, "/organizations/default")?.json_object()?;
    let mut fields = Map::new();
    fields.insert(
        "default".into(),
        organizations::organization(body.get("organization"))?,
    );
    Ok(fields)
}

/// The user fields of a recent run: triage never carries runtime details.
fn triage_run(run: Value) -> Value {
    let Value::Object(fields) = run else {
        return run;
    };
    Value::Object(
        fields
            .into_iter()
            .filter(|(key, _)| TRIAGE_RUN_FIELDS.contains(&key.as_str()))
            .collect(),
    )
}

fn runs_section(context: &mut Context<'_>) -> Result<Map<String, Value>, RunnerError> {
    let body = fetch(context, RUNS, "/runs")?.json_object()?;
    let recent = body
        .get("runs")
        .and_then(Value::as_array)
        .filter(|runs| runs.len() <= 200)
        .ok_or_else(|| protocol("the run list is missing or has more than 200 runs"))?
        .iter()
        .take(RECENT_RUNS)
        .map(|run| run_records::project_run(run, false).map(triage_run))
        .collect::<Result<Vec<_>, _>>()?;
    let live = recent
        .iter()
        .filter(|run| {
            run["status"]
                .as_str()
                .is_some_and(|status| LIVE_RUN_STATUSES.contains(&status))
        })
        .map(|run| run["run_id"].clone())
        .collect::<Vec<_>>();
    let mut fields = Map::new();
    fields.insert("recent".into(), Value::Array(recent));
    fields.insert("live".into(), Value::Array(live));
    Ok(fields)
}

fn jobs_section(context: &mut Context<'_>) -> Result<Map<String, Value>, RunnerError> {
    let body = fetch(context, JOBS, "/triggers")?.json_object()?;
    let (all, max) = jobs::project_jobs(&body)?;
    let shown = all
        .iter()
        .take(JOBS_SHOWN)
        .map(|entry| {
            let mut job = Map::new();
            for key in JOB_KEYS {
                if let Some(value) = entry.get(key).filter(|value| !value.is_null()) {
                    job.insert(key.into(), value.clone());
                }
            }
            Value::Object(job)
        })
        .collect::<Vec<_>>();
    let mut fields = Map::new();
    fields.insert("total".into(), json!(all.len()));
    fields.insert("max".into(), max);
    fields.insert("jobs".into(), Value::Array(shown));
    Ok(fields)
}

fn reason_of(problem: &Value) -> String {
    problem["details"]["reason"]
        .as_str()
        .or_else(|| problem["message"].as_str())
        .unwrap_or_default()
        .to_owned()
}

fn cents_text(cents: i64) -> String {
    format!("{}.{:02}", cents / 100, (cents % 100).abs())
}

/// `nextCommands`, derived from the sections in a fixed order: the
/// credential variable, an unreachable service, live runs, a low balance,
/// jobs reporting an error. Commands keep the output mode.
fn next_commands(context: &Context<'_>, result: &Value, reachable: bool) -> Vec<Value> {
    let mut next = Vec::new();
    let credential = &result["credential"];
    if !credential["problem"].is_null() {
        next.push(json!({
            "why": reason_of(&credential["problem"]),
            "argv": null,
            "env": credential["variable"],
        }));
    }
    if !reachable {
        next.push(json!({
            "why": "the service did not respond; check your network connection, then retry",
            "argv": context.follow_up_argv(&["service", "status"]),
            "env": null,
        }));
    }
    if let Some(live) = result["runs"]["live"].as_array() {
        let recent = result["runs"]["recent"]
            .as_array()
            .cloned()
            .unwrap_or_default();
        for run_id in live.iter().filter_map(Value::as_str).take(LIVE_WATCH_MAX) {
            let status = recent
                .iter()
                .find(|run| run["run_id"] == run_id)
                .and_then(|run| run["status"].as_str())
                .unwrap_or_default();
            next.push(json!({
                "why": format!("run {run_id} is {status}"),
                "argv": context.follow_up_argv(&["run", "watch", run_id]),
                "env": null,
            }));
        }
    }
    if result["wallet"]["low"] == true {
        let threshold = result["wallet"]["lowBelowCents"]
            .as_i64()
            .unwrap_or_default();
        let amount = threshold.to_string();
        next.push(json!({
            "why": format!(
                "the available balance ${} is below ${}",
                result["wallet"]["balance"]["available_dollars"].as_str().unwrap_or_default(),
                cents_text(threshold)
            ),
            "argv": context.follow_up_argv(&["wallet", "topup", "--amount-cents", &amount, "--preview"]),
            "env": null,
        }));
    }
    if let Some(shown) = result["jobs"]["jobs"].as_array() {
        for job in shown
            .iter()
            .filter(|job| job.get("last_error").is_some())
            .take(FAILING_JOBS_MAX)
        {
            let id = job["id"].as_str().unwrap_or_default();
            next.push(json!({
                "why": format!("job {id} reported an error"),
                "argv": context.follow_up_argv(&["job", "show", id]),
                "env": null,
            }));
        }
    }
    next
}

fn problem_line(label: &str, problem: &Value) -> String {
    format!(
        "{label}: {}: {}\n",
        problem["code"].as_str().unwrap_or_default(),
        human_safe_scalar(&reason_of(problem))
    )
}

fn text(value: &Value) -> String {
    human_safe_scalar(value.as_str().unwrap_or_default())
}

/// The human report: at most 24 lines (one per section, up to five runs,
/// five jobs and seven next commands).
fn human(result: &Value) -> String {
    let mut out = String::new();
    let health = &result["health"];
    if health["problem"].is_null() {
        let _ = writeln!(
            out,
            "Service: {}, default model {}",
            text(&health["status"]),
            text(&health["default_model"])
        );
    } else {
        out.push_str(&problem_line("Service", &health["problem"]));
    }
    let credential = &result["credential"];
    let variable = text(&credential["variable"]);
    let origin = match credential["source"].as_str() {
        Some("none") => format!("{variable} not set"),
        Some("environment") => format!("from {variable}"),
        Some("store") => "from the credential store".to_owned(),
        Some(source) => format!("{variable} from {source}"),
        None => variable,
    };
    let _ = write!(out, "Credential: {}, {origin}", text(&credential["state"]));
    if credential["problem"].is_null() {
        out.push('\n');
    } else {
        let _ = writeln!(
            out,
            "; {}",
            human_safe_scalar(&reason_of(&credential["problem"]))
        );
    }
    let wallet = &result["wallet"];
    if wallet.is_null() {
        out.push_str("Wallet: skipped\n");
    } else if wallet["problem"].is_null() {
        let balance = &wallet["balance"];
        let _ = write!(
            out,
            "Wallet: ${} available, ${} reserved",
            text(&balance["available_dollars"]),
            text(&balance["reserved_dollars"])
        );
        if wallet["low"] == true {
            let _ = write!(
                out,
                " (low: below ${})",
                cents_text(wallet["lowBelowCents"].as_i64().unwrap_or_default())
            );
        }
        out.push('\n');
    } else {
        out.push_str(&problem_line("Wallet", &wallet["problem"]));
    }
    let organization = &result["organization"];
    if organization.is_null() {
        out.push_str("Organization: skipped\n");
    } else if organization["problem"].is_null() {
        let default = &organization["default"];
        let _ = write!(out, "Organization: {}", text(&default["slug"]));
        if let Some(role) = default["role"].as_str() {
            let _ = write!(out, " ({})", human_safe_scalar(role));
        }
        out.push('\n');
    } else {
        out.push_str(&problem_line("Organization", &organization["problem"]));
    }
    let runs = &result["runs"];
    if runs.is_null() {
        out.push_str("Runs: skipped\n");
    } else if runs["problem"].is_null() {
        let recent = runs["recent"].as_array().cloned().unwrap_or_default();
        let live = runs["live"].as_array().map_or(0, Vec::len);
        let _ = writeln!(out, "Runs: {} recent, {live} live", recent.len());
        for run in &recent {
            let _ = writeln!(
                out,
                "  {}  {}  {}  {}",
                text(&run["run_id"]),
                text(&run["status"]),
                text(&run["model"]),
                text(&run["created_at"])
            );
        }
    } else {
        out.push_str(&problem_line("Runs", &runs["problem"]));
    }
    let jobs = &result["jobs"];
    if jobs.is_null() {
        out.push_str("Jobs: skipped\n");
    } else if jobs["problem"].is_null() {
        let total = jobs["total"].as_u64().unwrap_or_default();
        match jobs["max"].as_i64() {
            Some(max) => {
                let _ = writeln!(out, "Jobs: {total} of {max} allowed");
            }
            None => {
                let _ = writeln!(out, "Jobs: {total}");
            }
        }
        for job in jobs["jobs"].as_array().cloned().unwrap_or_default() {
            let name = job.get("name").map_or_else(|| "-".to_owned(), text);
            let _ = write!(
                out,
                "  {}  {}  {name}",
                text(&job["id"]),
                text(&job["type"])
            );
            if job.get("last_error").is_some() {
                out.push_str("  (error)");
            }
            out.push('\n');
        }
    } else {
        out.push_str(&problem_line("Jobs", &jobs["problem"]));
    }
    let next = result["nextCommands"]
        .as_array()
        .cloned()
        .unwrap_or_default();
    if next.is_empty() {
        out.push_str("Next: nothing needs attention\n");
    } else {
        out.push_str("Next:\n");
        for command in &next {
            let action = match command["argv"].as_array() {
                Some(argv) => render::argv_text(
                    &argv
                        .iter()
                        .filter_map(Value::as_str)
                        .map(str::to_owned)
                        .collect::<Vec<_>>(),
                ),
                None => format!("set {}", command["env"].as_str().unwrap_or_default()),
            };
            let _ = writeln!(
                out,
                "  {}  # {}",
                human_safe_scalar(&action),
                human_safe_scalar(command["why"].as_str().unwrap_or_default())
            );
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn cents_render_as_dollars() {
        assert_eq!(cents_text(500), "5.00");
        assert_eq!(cents_text(35), "0.35");
        assert_eq!(cents_text(12345), "123.45");
    }

    #[test]
    fn human_report_stays_within_25_lines_at_its_largest() {
        let run = json!({"run_id": "run_1", "status": "running", "model": "m", "created_at": "t"});
        let job = json!({"id": "j", "type": "schedule", "last_error": "x"});
        let command = json!({"why": "w", "argv": ["cli", "run", "watch", "run_1"], "env": null});
        let result = json!({
            "health": {"problem": null, "status": "ok", "default_model": "m", "models": ["m"]},
            "credential": {"variable": "V", "source": "environment", "state": "valid", "problem": null},
            "wallet": {"problem": null, "balance": {"available_dollars": "1.00", "reserved_dollars": "0.00"},
                       "low": true, "lowBelowCents": 500},
            "organization": {"problem": null, "default": {"slug": "o", "role": "admin"}},
            "runs": {"problem": null, "recent": vec![run; RECENT_RUNS], "live": ["run_1"]},
            "jobs": {"problem": null, "total": 9, "max": 10, "jobs": vec![job; JOBS_SHOWN]},
            "nextCommands": vec![command; LIVE_WATCH_MAX + 1 + FAILING_JOBS_MAX],
        });
        let text = human(&result);
        assert!(text.lines().count() <= 24);
        assert!(text.starts_with("Service: ok, default model m\n"), "{text}");
    }
}
